evennia/utils/evmenu.py

```diff
@@ -64,10 +64,41 @@ def another_node(caller, input_string, **kwargs):
 The return values must be given in the above order, but each can be
 returned as None as well. If the options are returned as None, the
 menu is immediately exited and the default "look" command is called.
-
-    text (str, tuple or None): Text shown at this node. If a tuple, the
-        second element in the tuple is a help text to display at this
-        node when the user enters the menu help command there.
+    text (str, tuple, dict or None): Text shown at this node. If a dict
+        it can also contain various formatting instructions. In such a
+        case the followng keys are already supported:
+            - 'text' (str): The text of the node.
+            - 'help' (str): The help text of the node.
+            - 'header' (str): Header text for the node.
+            - 'footer' (str): Footer text for the node.
+            - 'format' (str): String dictating the layout of the node if
+                the default layout is not being used. It currently
+                supports three options:
+                    - 'suppress': The node is not drawn at all
+                    - 'no_bars': The node is drawn similarly to the
+                        default format but without separator bars.
+                    - 'invert': The node is drawn similar to the
+                        default format but the options table will be
+                        above the text.
```

Comment:
I see what you are saying about the different formatters. I'm still not completely sold on these being necessities because if someone wants to do an inverted node with no bars they can subclass EvMenu, create a formatter function of ``invert_no_bars_format``, and override node_formatter (which is only 47 lines long now) to add 
``............elif nodetext['format'] == 'invert_no_bars':``
``................result = self.invert_no_bars_format(text, optionstext, header, footer)``

My goal wasn't to completely eliminate subclassing but to make it less necessary, easier to implement, and capable of handling things node-by-node without the user having to change the methodology.

With that said, I can still see validity to adding ``header_formatter``, ``text_formatter``, ``help_formatter``, ``options_formatter``, ``footer_formatter`` and ``node_formatter`` since it will expand the flexibility by a great deal without any subclassing. 

At present I think the only formatters for the subsections need to be ``bars`` or ``no_bars`` and then ``traditional``, ``invert`` or ``suppress`` for ``node_formatter``. Default behaviors when not called can be to create nodes in the current layout so that backwards compatibility is preserved.

How does that work for you?

Reply:
Here's an idea:

Instead of making independent ``node_formatter``, ``text_formatter``, etc. keys what if we make the ``formatter`` key part of the entry specific ``dict``? This would require us to create an additional key for ``contents`` for each of these but it would also keep things more organized.

Additionally, by biting the bullet and making ``return`` accept a ``dict`` instead of making the first part of the ``tuple`` accept a ``dict`` we can avoid something that has been mildly annoying me, mainly that the formatting instructions are separated from the options list.

Usage would look similar to this:
![Example](https://user-images.githubusercontent.com/8743461/82932433-4359c700-9f3d-11ea-9d56-55256fdcf332.png)

This would allow additional variables for custom formatters to easily be embedded by simply adding keys to the sub-dictionary. (Similar to how the ``node`` ``dict`` contains ``hidekeys`` and ``rows`` keys.

Backwards compatibility can be maintained by examining the return and seeing if it is a ``dict`` and just making standard assumptions if it is not.

Reply:
Overriding EvMenu is not quite the same thing, because we are talking about per-node formatting, something you can't currently easily reproduce with overriding. For changing the look of every node, then overriding is indeed the way to go. (This is also another advantage of formatting functions - you could just import and call the formatting function in your override if you wanted to - any such formatters we provide act as a library for people to use.)

I think returning a dict is not a bad idea; if expanding the `text` to return as a dict anyway, being able to include `options` in that return is a rather natural expansion. The original idea was to keep the node return simple, but it's clear there is a need for more complexity here. So this sounds like an ok change. And one can still keep a backwards-compatible handling.

Overall, the various sections of the return looks fine (rather CSS-like in a way). Having each section set aside with `formatter`, `content` etc does make it look quite clean, I like that. 

There is absolutely no reason to not allow `formatter` to accept a callable though - or, to follow the style used everywhere else, a tuple `(callable, {kwargs})`. This needs to be supported, there is no way around it. 

As for using strings like "no_bars" instead of callables, this may be ok if they are not hard-coded exceptions. That is, let the string-form be defined as being the name of a formatter-method on the EvMenu class. (so "no_bars" would mean using the method `EvMenu.format_no_bars`, "invert" would point a a method `EvMenu.format_invert` and so on). This would also make the string-based formatter style extendable - you could conceivably use the overriding of EvMenu to provide more such simple formatters.

Reply:
I've update the code so that formatters are string based . The function finally called is a modification of the original string (so for options 'no_bars' uses the ``options_format_no_bars`` function while for strings it would use the ``strings_format_no_bars`` function.